autoPyTorch/pipeline/components/training/trainer/AdversarialTrainer.py

```diff
@@ -0,0 +1,157 @@
+from copy import deepcopy
+import typing
+from typing import Dict, Optional, Tuple
+
+from ConfigSpace.configuration_space import ConfigurationSpace
+from ConfigSpace.hyperparameters import (
+    CategoricalHyperparameter,
+    UniformFloatHyperparameter,
+)
+
+import numpy as np
+
+import torch
+from torch.autograd import Variable
+from torch.utils.tensorboard.writer import SummaryWriter
+
+from autoPyTorch.constants import CLASSIFICATION_TASKS, STRING_TO_TASK_TYPES
+from autoPyTorch.pipeline.components.training.trainer.base_trainer import BaseTrainerComponent
+from autoPyTorch.utils.logging_ import PicklableClientLogger
+
+
+class AdversarialTrainer(BaseTrainerComponent):
+    def __init__(
+            self,
+            epsilon: float,
+            weighted_loss: bool = False,
+            random_state: typing.Optional[np.random.RandomState] = None,
+    ):
+        """
+        This class handles the training of a network for a single given epoch.
+
+        Args:
+            epsilon (float): The perturbation magnitude.
+
+        """
+        super().__init__(random_state=random_state)
+        self.epsilon = epsilon
+        self.weighted_loss = weighted_loss
+
+    def data_preparation(self, X: np.ndarray, y: np.ndarray,
+                         ) -> typing.Tuple[typing.Tuple[np.ndarray, np.ndarray], typing.Dict[str, np.ndarray]]:
+        """Generate adversarial examples from the original inputs.
+
+        Args:
+            X (np.ndarray): The batch training features
+            y (np.ndarray): The batch training labels
+
+        Returns:
+            typing.Tuple[np.ndarray, np.ndarray]: original examples, adversarial examples.
+            typing.Dict[str, np.ndarray]: arguments to the criterion function.
+        """
+        X_adversarial = self.fgsm_attack(X, y)
+        return (X, X_adversarial), {'y_a': y}
+
+    def criterion_preparation(self, y_a: np.ndarray, y_b: np.ndarray = None, lam: float = 1.0
+                              ) -> typing.Callable:
+        # Initial implementation, consider the adversarial loss and the normal network loss
+        # equally.
+        return lambda criterion, pred, adversarial_pred: 0.5 * criterion(pred, y_a) + \
```

Comment:
does it need to be fixed? maybe we can sample randomly or set lam: 0.5 in data_preparation

Reply:
That is a really good point. I made it fixed since that is how it was in the paper and like that it cares equally about both. 
It makes sense for it to be a hyperparameter too though.

Reply:
okay then we can add a hyperparameter later, when we merge this into refactor_development

Reply:
Yep, that was exactly what I was thinking.